Update `expr` matcher for Edition 2024 and add `expr_2021` nonterminal

eholk opened the pull request:
This commit adds a new nonterminal `expr_2021` in macro patterns, and `expr_fragment_specifier_2024` feature flag.

This change also updates `expr` so that on Edition 2024 it will also match `const { ... }` blocks, while `expr_2021` preserves the current behavior of `expr`, matching expressions without `const` blocks.

Joint work with @vincenzopalazzo.

Issue #123742

eholk committed:
Add expr_2021 nonterminal and feature flag

This commit adds a new nonterminal `expr_2021` in macro patterns, and
`expr_fragment_specifier_2024` feature flag. For now, `expr` and
`expr_2021` are treated the same, but in future PRs we will update
`expr` to match to new grammar.

Co-authored-by: Vincezo Palazzo <vincenzopalazzodev@gmail.com>

eholk committed:
Add test that expr_2021 only works on 2024 edition

Co-authored-by: Vincenzo Palazzo <vincenzopalazzodev@gmail.com>

eholk commented on compiler/rustc_ast/src/token.rs:
```diff
@@ -905,6 +907,7 @@ impl NonterminalKind {
             },
             sym::pat_param => NonterminalKind::PatParam { inferred: false },
             sym::expr => NonterminalKind::Expr,
+            sym::expr_2021 if edition() >= Edition::Edition2024 => NonterminalKind::Expr2021,
```
From [the RFC](https://github.com/rust-lang/rfcs/blob/master/text/3531-macro-fragment-policy.md#:~:text=In%20the%20current%20edition%2C%20the%20next%20edition%2C%20and%20as%20many%20other%20editions%20as%20practical%2C%20add%20a%20new%20fragment%20specifier%20that%20preserves%20the%20behavior%20of%20the%20existing%20fragment%20specifier.):

> In the current edition, the next edition, and as many other editions as practical, add a new fragment specifier that preserves the behavior of the existing fragment specifier.

My interpretation of this is that we need to allow `expr_2021` on Edition 2021, and if practical also support it on Edition 2018 and 2015. I don't see any reason why we shouldn't be able to support this on all editions.

eholk replied:
Actually, we probably can't do 2015 because `async` wasn't a keyword then.

vincenzopalazzo replied:
**TL;DR:** It's safe to assume that `expr_2021` will work for every previous edition until we encounter another `expr_xx`, such as `expr_2015` (if this were the case).

One practical idea to support this range of editions is that `expr_2021` remains backwards compatible until `expr_xx`. If there is another `expr_xx`, then the compatibility range includes everything.

However, consider if someone uses the `async` keyword in an edition like 2015; the code will likely complain that `async` is not supported, right? Perhaps not, if the keyword is used only inside the macros.

eholk replied:
This is fixed, as of [e0b7d92](https://github.com/rust-lang/rust/pull/123865/commits/e0b7d923b4900b382a2f20abb75a5aa000284542) we allow `expr_2021` on Edition 2021.

eholk committed:
expr_2021 should be allowed on edition 2021 and later

vincenzopalazzo committed:
Macros: match const { ... } with expr nonterminal in edition 2024

Co-authored-by: Eric Holk <eric@theincredibleholk.org>
Signed-off-by: Vincenzo Palazzo <vincenzopalazzodev@gmail.com>

fmease commented on tests/ui/macros/expr_2021_inline_const.rs:
I wonder if there are any crazy edge cases to consider like nested MBEs where the innermost fragment specifier is a macro metavariable and decl macros come from different crates. E.g:

Upstream:

```rs
//@ edition: ???
#![feature(macro_metavar_expr)]

#[macro_export]
macro_rules! outer {
    ($frag:ident) => {
        macro_rules! inner {
            ($$it:$frag) => {}
        }
    }
}
```

Downstream:

```rs
//@ edition: ???

upstream::outer!(expr);
inner!(const { 0 });
```

I don't know if this yields anything interesting. I'm a bit tired right now, it's late. Just wanted to mention it.

eholk replied:
I've been tying my brain in knots about these kinds of scenarios recently, but I don't have good answers for them.

I think my approach would be to go ahead and merge this and then get wg-macros to start creating test cases to make sure what the behavior we have is and whether that's the behavior we want.

eholk replied:
I added a bullet about this to #123742 so we don't lose the suggestion.

eholk replied:
We have a PR with that adds a test case around this scenario: https://github.com/rust-lang/rust/pull/129755

Not entirely sure what the correct behavior should be, but it seems like the behavior we have now isn't bad.

fmease commented:
I think the biggest question left is what exactly should be included in Rust 2024's `expr` as you've started to discuss in the tracking issue.

eholk committed:
Apply code review suggestions

- use feature_err to report unstable expr_2021
- Update downlevel expr_2021 diagnostics

Co-authored-by: León Orell Valerian Liehr <me@fmease.dev>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/8902963757/job/24449860423) [(plain)](https://github.com/rust-lang/rust/commit/0d434375f487ed17fac91be783077f44658d7775/checks/24449860423/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#16 exporting to docker image format
#16 sending tarball 29.5s done
#16 DONE 34.4s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
.................................................i......................................  8976/16796
........................................................................................  9064/16796
........................................................................................  9152/16796
......................i.....
[ui] tests/ui/macros/expr_2021_inline_const.rs#edi2024 ... F

[ui] tests/ui/macros/expr_2021_inline_const.rs#edi2021 ... F
...............ii..................................................................i....  9328/16796
........................................................................................  9416/16796
........................................................................................  9504/16796
.......................ii...............................................................  9592/16796
---
............................................................................

failures:

---- [ui] tests/ui/macros/expr_2021_inline_const.rs#edi2024 stdout ----


13 LL |     ($e:expr_2021) => {
15 
- error: aborting due to 1 previous error
+ warning: the feature `inline_const` has been stable since 1.80.0-nightly and no longer requires an attribute to enable
+   --> $DIR/expr_2021_inline_const.rs:7:12
---
18 


The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/macros/expr_2021_inline_const.edi2024/expr_2021_inline_const.edi2024.stderr
To only update this specific test, also pass `--test-args macros/expr_2021_inline_const.rs`


error in revision `edi2024`: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/macros/expr_2021_inline_const.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--cfg" "edi2024" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/macros/expr_2021_inline_const.edi2024" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/macros/expr_2021_inline_const.edi2024/auxiliary" "--edition=2024" "-Z" "unstable-options"
--- stderr -------------------------------
error: no rules expected the token `const`
##[error]  --> /checkout/tests/ui/macros/expr_2021_inline_const.rs:22:12
   |
   |
LL | macro_rules! m2021 {
   | ------------------ when calling this macro
...
LL |     m2021!(const { 1 }); //~ ERROR: no rules expected the token `const`
   |            ^^^^^ no rules expected this token in macro call
   |
note: while trying to match meta-variable `$e:expr_2021`
  --> /checkout/tests/ui/macros/expr_2021_inline_const.rs:11:6
   |
LL |     ($e:expr_2021) => {

warning: the feature `inline_const` has been stable since 1.80.0-nightly and no longer requires an attribute to enable
##[warning]  --> /checkout/tests/ui/macros/expr_2021_inline_const.rs:7:12
   |
---
error: aborting due to 1 previous error; 1 warning emitted
------------------------------------------


---- [ui] tests/ui/macros/expr_2021_inline_const.rs#edi2021 stdout ----

28 LL |     ($e:expr) => {
29    |      ^^^^^^^
30 
---
33 


The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/macros/expr_2021_inline_const.edi2021/expr_2021_inline_const.edi2021.stderr
To only update this specific test, also pass `--test-args macros/expr_2021_inline_const.rs`


error in revision `edi2021`: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/macros/expr_2021_inline_const.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--cfg" "edi2021" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/macros/expr_2021_inline_const.edi2021" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/macros/expr_2021_inline_const.edi2021/auxiliary" "--edition=2021"
--- stderr -------------------------------
error: no rules expected the token `const`
##[error]  --> /checkout/tests/ui/macros/expr_2021_inline_const.rs:22:12
   |
   |
LL | macro_rules! m2021 {
   | ------------------ when calling this macro
...
LL |     m2021!(const { 1 }); //~ ERROR: no rules expected the token `const`
   |            ^^^^^ no rules expected this token in macro call
   |
note: while trying to match meta-variable `$e:expr_2021`
  --> /checkout/tests/ui/macros/expr_2021_inline_const.rs:11:6
   |
LL |     ($e:expr_2021) => {

error: no rules expected the token `const`
##[error]  --> /checkout/tests/ui/macros/expr_2021_inline_const.rs:23:12
   |
   |
LL | macro_rules! m2024 {
   | ------------------ when calling this macro
...
LL |     m2024!(const { 1 }); //[edi2021]~ ERROR: no rules expected the token `const`
   |            ^^^^^ no rules expected this token in macro call
   |
note: while trying to match meta-variable `$e:expr`
  --> /checkout/tests/ui/macros/expr_2021_inline_const.rs:17:6
LL |     ($e:expr) => {
   |      ^^^^^^^

warning: the feature `inline_const` has been stable since 1.80.0-nightly and no longer requires an attribute to enable
```

</details>

eholk commented on compiler/rustc_expand/src/mbe/quoted.rs:
```diff
@@ -92,6 +92,12 @@ pub(super) fn parse(
+                                        && !features.expr_fragment_specifier_2024
+                                    {
+                                        sess.dcx()
+                                            .emit_err(errors::Expr2021IsExperimental { span });
```
Thanks for the suggestion. This should be done now.

eholk commented on tests/ui/macros/expr_2021_old_edition.stderr:
```diff
@@ -0,0 +1,25 @@
+error: invalid fragment specifier `expr_2021`
```
I updated the help text so it should be better now. The way I did it was pretty clumsy though, so I'm open to suggestions for how to improve it.

eholk commented:
@fmease - I addressed your simpler comments, so I think this should be ready for another look.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/8913512530/job/24479146252) [(plain)](https://github.com/rust-lang/rust/commit/acbc58c41d3b0dab4f0a29f4afafde3f3f30f320/checks/24479146252/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#17 exporting to docker image format
#17 sending tarball 29.7s done
#17 DONE 38.4s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
..................................................i.....................................  8976/16797
........................................................................................  9064/16797
........................................................................................  9152/16797
.......................i....
[ui] tests/ui/macros/expr_2021_inline_const.rs#edi2021 ... F
.....
[ui] tests/ui/macros/expr_2021_inline_const.rs#edi2024 ... F
................ii..................................................................i...  9328/16797
........................................................................................  9416/16797
........................................................................................  9504/16797
........................ii..............................................................  9592/16797
---
.............................................................................

failures:

---- [ui] tests/ui/macros/expr_2021_inline_const.rs#edi2021 stdout ----

28 LL |     ($e:expr) => {
29    |      ^^^^^^^
30 
---
33 


The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/macros/expr_2021_inline_const.edi2021/expr_2021_inline_const.edi2021.stderr
To only update this specific test, also pass `--test-args macros/expr_2021_inline_const.rs`


error in revision `edi2021`: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/macros/expr_2021_inline_const.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--cfg" "edi2021" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/macros/expr_2021_inline_const.edi2021" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/macros/expr_2021_inline_const.edi2021/auxiliary" "--edition=2021"
--- stderr -------------------------------
error: no rules expected the token `const`
##[error]  --> /checkout/tests/ui/macros/expr_2021_inline_const.rs:22:12
   |
   |
LL | macro_rules! m2021 {
   | ------------------ when calling this macro
...
LL |     m2021!(const { 1 }); //~ ERROR: no rules expected the token `const`
   |            ^^^^^ no rules expected this token in macro call
   |
note: while trying to match meta-variable `$e:expr_2021`
  --> /checkout/tests/ui/macros/expr_2021_inline_const.rs:11:6
   |
LL |     ($e:expr_2021) => {

error: no rules expected the token `const`
##[error]  --> /checkout/tests/ui/macros/expr_2021_inline_const.rs:23:12
   |
   |
LL | macro_rules! m2024 {
   | ------------------ when calling this macro
...
LL |     m2024!(const { 1 }); //[edi2021]~ ERROR: no rules expected the token `const`
   |            ^^^^^ no rules expected this token in macro call
   |
note: while trying to match meta-variable `$e:expr`
  --> /checkout/tests/ui/macros/expr_2021_inline_const.rs:17:6
LL |     ($e:expr) => {
   |      ^^^^^^^

warning: the feature `inline_const` has been stable since 1.80.0-nightly and no longer requires an attribute to enable
---
error: aborting due to 2 previous errors; 1 warning emitted
------------------------------------------


---- [ui] tests/ui/macros/expr_2021_inline_const.rs#edi2024 stdout ----


13 LL |     ($e:expr_2021) => {
15 
- error: aborting due to 1 previous error
+ warning: the feature `inline_const` has been stable since 1.80.0-nightly and no longer requires an attribute to enable
+   --> $DIR/expr_2021_inline_const.rs:7:12
---
18 


The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/macros/expr_2021_inline_const.edi2024/expr_2021_inline_const.edi2024.stderr
To only update this specific test, also pass `--test-args macros/expr_2021_inline_const.rs`


error in revision `edi2024`: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/macros/expr_2021_inline_const.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--cfg" "edi2024" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/macros/expr_2021_inline_const.edi2024" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/macros/expr_2021_inline_const.edi2024/auxiliary" "--edition=2024" "-Z" "unstable-options"
--- stderr -------------------------------
error: no rules expected the token `const`
##[error]  --> /checkout/tests/ui/macros/expr_2021_inline_const.rs:22:12
   |
   |
LL | macro_rules! m2021 {
   | ------------------ when calling this macro
...
LL |     m2021!(const { 1 }); //~ ERROR: no rules expected the token `const`
   |            ^^^^^ no rules expected this token in macro call
   |
note: while trying to match meta-variable `$e:expr_2021`
  --> /checkout/tests/ui/macros/expr_2021_inline_const.rs:11:6
   |
LL |     ($e:expr_2021) => {

warning: the feature `inline_const` has been stable since 1.80.0-nightly and no longer requires an attribute to enable
##[warning]  --> /checkout/tests/ui/macros/expr_2021_inline_const.rs:7:12
   |
```

</details>

vincenzopalazzo commented:
Looks like we should update the tests

```
  failures:
  
  ---- [ui] tests/ui/macros/expr_2021_inline_const.rs#edi2021 stdout ----
  diff of stderr:
  
  28	LL |     ($e:expr) => {
  29	   |      ^^^^^^^
  30	
  -	error: aborting due to 2 previous errors
  +	warning: the feature `inline_const` has been stable since 1.80.0-nightly and no longer requires an attribute to enable
  +	  --> $DIR/expr_2021_inline_const.rs:7:12
  +	   |
  +	LL | #![feature(inline_const)]
  +	   |            ^^^^^^^^^^^^
  +	   |
  +	   = note: `#[warn(stable_features)]` on by default
  +	
  +	error: aborting due to 2 previous errors; 1 warning emitted
  32	
  33	
```

fmease commented on tests/ui/macros/expr_2021_old_edition.stderr:
```diff
@@ -0,0 +1,26 @@
+LL |     ($e:expr_2021) => {
+   |      ^^^^^^^^^^^^
+   |
+   = help: fragment specifier `expr_2021` requires Rust 2021 or later
+           valid fragment specifiers are `ident`, `block`, `stmt`, `expr`, `pat`, `ty`, `lifetime`, `literal`, `path`, `meta`, `tt`, `item` and `vis`
```
I would probably split this into two separate `help` messages but er, doesn't matter

eholk replied:
Yeah, I think that'd be better. Let me poke around at it some more, I'm still getting the hang of the diagnostics system.

fmease commented on tests/ui/macros/expr_2021_inline_const.edi2021.stderr:
This (intentional) behavior is a bit unfortunate but it makes sense. Well, it makes the most sense when edition-dependent keywords are involved.

fmease commented on compiler/rustc_parse/src/parser/nonterminal.rs:
```diff
@@ -37,13 +38,19 @@ impl<'a> Parser<'a> {
+                token.can_begin_expr()
+                // This exception is here for backwards compatibility.
+                && !token.is_keyword(kw::Let)
+                && (token.span.edition() >= Edition::Edition2024 || !token.is_keyword(kw::Const))
```
not applied? :P

eholk replied:
Oh weird, I thought I had done this one.

fmease reviewed (approved):
Thanks! My suggestion https://github.com/rust-lang/rust/pull/123865/files#r1577167279 wasn't applied, I assume that's an oversight because you've applied suggestions of mine that are similar to this.

Should be good to go after that, work like double-checking the behavior of weird nested macro decls can be done in follow-ups as suggested by you.

r=me then.

eholk committed:
Update compiler/rustc_parse/src/parser/nonterminal.rs

Co-authored-by: León Orell Valerian Liehr <me@fmease.dev>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9132660108/job/25114466600) [(plain)](https://github.com/rust-lang/rust/commit/c577e9e70254a21b4eae4a750d6aad0b32ff54a9/checks/25114466600/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#16 exporting to docker image format
#16 sending tarball 29.7s done
#16 DONE 32.3s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
```

</details>

eholk commented:
@bors r=@fmease

bors commented:
:pushpin: Commit 3986ea0ea555ab6de8af46932da94c26aa1c28ee has been approved by `fmease`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=fmease 3986ea0ea555ab6de8af46932da94c26aa1c28ee -->
<!-- homu: {"type":"Approved","sha":"3986ea0ea555ab6de8af46932da94c26aa1c28ee","approver":"fmease","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 3986ea0ea555ab6de8af46932da94c26aa1c28ee with merge 9b75a4388143a163b77fa7d458e4aa4dd34ac1bd...
<!-- homu: {"type":"BuildStarted","head_sha":"3986ea0ea555ab6de8af46932da94c26aa1c28ee","merge_sha":"9b75a4388143a163b77fa7d458e4aa4dd34ac1bd"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/9134676483/job/25123166391)
Approved by: fmease
Pushing 9b75a4388143a163b77fa7d458e4aa4dd34ac1bd to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"fmease","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/9134676483/job/25123166391"},"merge_sha":"9b75a4388143a163b77fa7d458e4aa4dd34ac1bd"} -->

rust-timer commented:
Finished benchmarking commit (9b75a4388143a163b77fa7d458e4aa4dd34ac1bd): [comparison URL](https://perf.rust-lang.org/compare.html?start=1a7397988684934ae01a71f524bdfff24895d8cc&end=9b75a4388143a163b77fa7d458e4aa4dd34ac1bd&stat=instructions:u).

### Overall result: ❌ regressions - ACTION NEEDED


**Next Steps**: If you can justify the regressions found in this perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please open an issue or create a new PR that fixes the regressions, add a comment linking to the newly created issue or PR, and then add the `perf-regression-triaged` label to this PR.

@rustbot label: +perf-regression
cc @rust-lang/wg-compiler-performance



### [Instruction count](https://perf.rust-lang.org/compare.html?start=1a7397988684934ae01a71f524bdfff24895d8cc&end=9b75a4388143a163b77fa7d458e4aa4dd34ac1bd&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.2% | [0.2%, 0.3%] | 5     |
| Regressions ❌ <br /> (secondary)  | 0.5% | [0.2%, 1.7%] | 7     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 0.2% | [0.2%, 0.3%] | 5     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=1a7397988684934ae01a71f524bdfff24895d8cc&end=9b75a4388143a163b77fa7d458e4aa4dd34ac1bd&stat=max-rss)
<details>
<summary>Results (primary 0.0%, secondary -4.7%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 2.6%  | [2.6%, 2.6%]   | 1     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -1.3% | [-2.3%, -0.2%] | 2     |
| Improvements ✅ <br /> (secondary) | -4.7% | [-4.7%, -4.7%] | 1     |
| All ❌✅ (primary)                 | 0.0%  | [-2.3%, 2.6%]  | 3     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=1a7397988684934ae01a71f524bdfff24895d8cc&end=9b75a4388143a163b77fa7d458e4aa4dd34ac1bd&stat=cycles:u)
<details>
<summary>Results (primary -3.1%, secondary 2.5%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 4.1%  | [2.0%, 8.1%]   | 8     |
| Improvements ✅ <br /> (primary)   | -3.1% | [-3.1%, -3.1%] | 1     |
| Improvements ✅ <br /> (secondary) | -1.7% | [-2.1%, -1.3%] | 3     |
| All ❌✅ (primary)                 | -3.1% | [-3.1%, -3.1%] | 1     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=1a7397988684934ae01a71f524bdfff24895d8cc&end=9b75a4388143a163b77fa7d458e4aa4dd34ac1bd&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 668.756s -> 668.325s (-0.06%)
**Artifact size**: 316.02 MiB -> 316.15 MiB (0.04%)

rylev commented:
@eholk @vincenzopalazzo @fmease the regression seems legit. It seems the primary benchmark affected by this `html5ever` seems to spend lots of time in `expand_crate` which could be impacted by this change. Thoughts on whether this needs further investigation or if we should just treat that particular benchmark as a quasi-stress-test and move on?

vincenzopalazzo commented:
@rylev I was planning to discuss with @eholk and take a look at the code, idk if @eholk already gave a look at the code, but I need a couple of more days

fmease commented:
Update: In #125389 I've checked the perf of a full revert. Doing a naive inversion / deduction, this PR "might've only had" 8 counts of secondary regressions (mean: +0.3%, range: [+0.5%, +0.3%]). Anyways, if I find the time I will look deeper into this. Otherwise, I'll just wait for https://github.com/rust-lang/rust/pull/123865#issuecomment-2122672475.

vincenzopalazzo commented on compiler/rustc_parse/src/parser/nonterminal.rs:
```diff
@@ -37,13 +37,19 @@ impl<'a> Parser<'a> {
+                token.can_begin_expr()
+                // This exception is here for backwards compatibility.
+                && !token.is_keyword(kw::Let)
+                && (token.span.edition().at_least_rust_2024() || !token.is_keyword(kw::Const))
```
I am not a performance expert, so I might be wrong here, but I think the only thing that could cause a performance issue is the token.span.edition().

If we follow the call stack, the edition() will resolve in the following call: https://github.com/rust-lang/rust/blob/master/compiler/rustc_span/src/hygiene.rs#L877-L879.

If we repeat this call for every expression token, we could see a performance decrease.

vincenzopalazzo reviewed (commented):
I finally had a look at the code and I post my analysis, I am not sure how to solve this BTW :/
